File: Arrays/Algorithms/Sorting/Non-Comparison/bucket_sort.py
def bucket_sort(arr):
	n = len(arr)
	buckets = [[] for _ in range(n)]
	for x in arr:
		idx = int(n * x)
		buckets[idx].append(x)
	for bucket in buckets:
		bucket.sort()
	idx = 0
	for bucket in buckets:
		for v in bucket:
			arr[idx] = v
			idx += 1

if __name__ == "__main__":
	n = int(input())
	arr = list(map(float, input().split()))
	bucket_sort(arr)
	print(*[f"{x:.5f}" for x in arr])

File: Arrays/Algorithms/Sorting/Non-Comparison/bucket_sort.cpp
#include <bits/stdc++.h>
using namespace std;

void bucketSort(vector<float>& arr) {
	int n = arr.size();
	vector<vector<float>> buckets(n);
	for (int i = 0; i < n; i++) {
		int idx = n * arr[i];
		buckets[idx].push_back(arr[i]);
	}
	for (int i = 0; i < n; i++) sort(buckets[i].begin(), buckets[i].end());
	int idx = 0;
	for (int i = 0; i < n; i++) {
		for (float v : buckets[i]) arr[idx++] = v;
	}
}

int main() {
	int n;
	cin >> n;
	vector<float> arr(n);
	for (int i = 0; i < n; i++) cin >> arr[i];
	bucketSort(arr);
	for (float x : arr) cout << fixed << setprecision(5) << x << ' ';
	cout << endl;
	return 0;
}
